arch/xtensa/src/esp32/esp32_spi.c

```diff
@@ -0,0 +1,1076 @@
+#include <errno.h>
+#include <debug.h>
+#include <time.h>
+
+#include <nuttx/arch.h>
+#include <nuttx/irq.h>
+#include <nuttx/clock.h>
+#include <nuttx/semaphore.h>
+#include <nuttx/spi/spi.h>
+
+#include <arch/board/board.h>
+
+#include "esp32_spi.h"
+#include "esp32_gpio.h"
+#include "esp32_cpuint.h"
+
+#include "xtensa.h"
+#include "hardware/esp32_gpio_sigmap.h"
+#include "hardware/esp32_dport.h"
+#include "hardware/esp32_spi.h"
+#include "hardware/esp32_soc.h"
+#include "rom/esp32_gpio.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* SPI Default speed (limited by clock divider) */
+
+#define SPI_FREQ_DEFAULT  400000
+
+/* SPI Device hardware configuration */
+
+struct esp32_spi_config_s
+{
+  uint32_t reg_base;          /* SPI register base address */
+
+  uint32_t clk_freq;          /* SPI clock frequency */
+  enum spi_mode_e mode;       /* SPI default mode */
+
+  uint8_t cs_pin;             /* GPIO configuration for CS */
+  uint8_t mosi_pin;           /* GPIO configuration for MOSI */
+  uint8_t miso_pin;           /* GPIO configuration for MISO */
+  uint8_t clk_pin;            /* GPIO configuration for CLK */
+
+  uint8_t cpu;                /* CPU ID */
+  uint8_t periph;             /* peripher ID */
+  uint8_t irq;                /* Interrupt ID */
+
+  uint32_t clk_bit;           /* Clock enable bit */
+  uint32_t rst_bit;           /* SPI reset bit */
+
+  uint32_t cs_insig;          /* SPI CS input signal index */
+  uint32_t cs_outsig;         /* SPI CS output signal index */
+  uint32_t mosi_insig;        /* SPI MOSI input signal index */
+  uint32_t mosi_outsig;       /* SPI MOSI output signal index */
+  uint32_t miso_insig;        /* SPI MISO input signal index */
+  uint32_t miso_outsig;       /* SPI MISO output signal index */
+  uint32_t clk_insig;         /* SPI CLK input signal index */
+  uint32_t clk_outsig;        /* SPI CLK output signal index */
+};
+
+struct esp32_spi_priv_s
+{
+  /* Externally visible part of the SPI interface */
+
+  struct spi_dev_s spi_dev;
+
+  /* Port configuration */
+
+  const struct esp32_spi_config_s *config;
+
+  int refs;                    /* Referernce count */
+
+  /* Held while chip is selected for mutual exclusion */
+
+  sem_t            exclsem;
+
+  uint32_t         frequency;   /* Requested clock frequency */
+  uint32_t         actual;      /* Actual clock frequency */
+
+  enum spi_mode_e  mode;        /* Actual SPI hardware mode */
+
+  /* Actual SPI send/receive bits once transmission */
+
+  uint8_t          nbits;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int esp32_spi_lock(FAR struct spi_dev_s *dev, bool lock);
+static void esp32_spi_select(FAR struct spi_dev_s *dev,
```

Comment:
This is also prototyped in the header file and implemented in board logic.

Reply:
I think only spi_select should be in the header, at least on stm32 and nrf52 I only found _select, not _lock.

Reply:
Yes, I was referring to spi_select not _lock.